src/dust/coagulation/coagulation.hpp

```diff
@@ -0,0 +1,968 @@
+//! \fn  void Dust::Coagulation::ComputeError
+//  \brief
+KOKKOS_INLINE_FUNCTION
+Real ComputeError(parthenon::team_mbr_t const &mbr, const int &mimax, const int &mimax2,
+                  const Real &h, const Real &h0, const ScratchPad1D<Real> &dustdens,
+                  const ScratchPad1D<Real> &source, const ScratchPad1D<Real> &nQs,
+                  const Real &err_eps) {
+  Real errmax = 0.0;
+  parthenon::par_reduce_inner(
+      parthenon::inner_loop_pattern_ttr_tag, mbr, 0, std::min(mimax, mimax2),
+      [&](const int i, Real &lmax) {
+        const Real dscale = std::abs(dustdens(i)) + std::abs(h0 * source(i));
+        lmax = std::max(lmax, std::abs(0.5 * h * (nQs(i) - source(i)) / dscale));
+      },
+      Kokkos::Max<Real>(errmax));
+  return errmax / err_eps;
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn  void Dust::Coagulation::
+//  \brief
+KOKKOS_INLINE_FUNCTION
+void Source(const parthenon::team_mbr_t &mbr, const int &nm1, const int &mimax,
+            const int &pgrid, const ScratchPad1D<Real> &source,
+            const ScratchPad1D<Real> &dustdens, const ScratchPad1D<Real> &vel,
+            const ScratchPad1D<Real> &stime, const StateParams &kernel,
+            const ParArray2D<int> &idx_largest, const ParArray1D<Real> &mass_grid,
+            const ParArray3D<Real> &coagR3D, const ParArray3D<int> &Kijk_sym_ind,
+            const ParArray3D<Real> &Kijk_sym, const bool &surface,
+            const RateParams &rate) {
+  ZeroSource(mbr, nm1, source);
+  InitializeSource(mbr, nm1, mimax, source, dustdens, vel, stime, kernel, mass_grid,
+                   coagR3D, Kijk_sym_ind, Kijk_sym, surface, rate);
+  FragmentationSource(mbr, nm1, mimax, source, dustdens, vel, stime, kernel, idx_largest,
+                      mass_grid, coagR3D, surface, rate);
+  CrateringSource(mbr, nm1, mimax, source, dustdens, vel, stime, kernel, mass_grid,
+                  coagR3D, surface, rate);
+  FinalizeSource(mbr, nm1, mimax, pgrid, source, dustdens, vel, stime, kernel, mass_grid,
+                 coagR3D, surface, rate);
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn  void Dust::Coagulation::
+//  \brief
+KOKKOS_INLINE_FUNCTION
+void SourceNQ(const parthenon::team_mbr_t &mbr, const int &n, const int &nm1,
+              const int &mimax, const int &pgrid, const ScratchPad1D<Real> &Q,
+              const ScratchPad1D<Real> &nQs, const ScratchPad1D<Real> &dustdens,
+              const ScratchPad1D<Real> &vel, const ScratchPad1D<Real> &stime,
+              const StateParams &kernel, const ParArray1D<Real> &mass_grid,
+              const ParArray3D<Real> &coagR3D, const ParArray3D<int> &Kijk_sym_ind,
+              const ParArray3D<Real> &Kijk_sym, const Real &chi, const bool &surface,
+              const RateParams &rate) {
+  InitializeSourceNQ(mbr, nm1, mimax, Q, nQs, dustdens, vel, stime, kernel, mass_grid,
```

Comment:
Just curious - why does this `SourceNQ` not have a `ZeroSourceNQ`? Otherwise it looks analogous to the `Source` routine above (sorry if missing something - don't know this code/physics really)